File: src/main/java/Electro2D/Protease.java
package main.java.Electro2D; /**
 * @author Amanda Fisher
 */

import java.util.ArrayList;

/**
 * Contains the methods needed for a Protease who cuts before or after apecific proteins without any other conditions.  Otherwise the method cut needs to be overwritten and new conditions added
 */
public abstract class Protease {

    final private char[] cuts;
    final private boolean cutsBefore;// does it cut before or after the proteins in cutAminoAcids

    public Protease(char[] cuts, boolean cutsBefore) {
    	this.cuts = cuts;
        this.cutsBefore = cutsBefore;
	}

	/**
	 * Cut array list.
	 *
	 * @param sequence the sequence
	 * @return the array list
	 * @throws ProteaseException the protease exception
	 */
	public ArrayList<String> cut(String sequence) throws ProteaseException {
		ArrayList<String> cutSequences = new ArrayList<>();
		sequence = checkSequence(sequence);
		// runs the sequence through the sequence checker method in order to detect any
		// errors
		char[] chars = sequence.toCharArray();
		int nCut = cuts.length;
		int p0 = 0, p1 = 0;
		for (int i = 0, n = chars.length - 1; i <= n; i++) {
			char aa = chars[i];
			if (!cutsBefore) {
				p1++;
			}
			char afterAA = (i < n ? chars[i + 1] : ' ');
			for (int j = 0; j < nCut; j++) {
				char c = cuts[j];
				if (shouldCutHere(aa, c, afterAA)) {
					makeIon(sequence, p0, p1, cutSequences);
					p0 = p1;
				}
			}
			if (cutsBefore) {
				p1++;
			}
		}
		if (p1 < sequence.length()) {
			// added by Bob Hanson - right? Do you get the final ion as well?
			makeIon(sequence, p1, sequence.length(), cutSequences);
		}
		return cutSequences;
	}

    protected boolean shouldCutHere(char aa, char c, char afterAA) {
    	// overridden only by Trypsin
    	return (aa == c);
    }

    private static String checkSequence(String sequence) throws ProteaseException {
        if (sequence.contains(" ")) {
            throw new ProteaseException("Sequence to be cut must not contain spaces.");
        } else if (sequence.matches(".*\\d.*")) {
            throw new ProteaseException("Sequence to be cut must not contain numbers.");
        } else if (sequence.matches(".*[a-z].*")) {
            throw new ProteaseException("Sequence to be cut must contain all upper case letters.");
        }
        return sequence;
    }

	/**
	 * Takes the characters collected by cut and turns them in to a string that
	 * represents the sequence of the Ion
	 *
	 * @param buildingIons The current ion
	 * @param cutSequence  The sequence to be cut
	 */
	private static void makeIon(String sequence, int p0, int p1, ArrayList<String> cutSequence) {
		if (p0 < p1)
			cutSequence.add(sequence.substring(p0, p1));
	}


}

File: site/swingjs/j2s/main/java/Electro2D/Protease.js
(function(){var P$=Clazz.newPackage("main.java.Electro2D"),I$=[[0,'java.util.ArrayList']],I$0=I$[0],$I$=function(i,n){return((i=(I$[i]||(I$[i]=Clazz.load(I$0[i])))),!n&&i.$load$&&Clazz.load(i,2),i)};
/*c*/var C$=Clazz.newClass(P$, "Protease");

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
},1);

C$.$fields$=[['Z',['cutsBefore'],'O',['cuts','char[]']]]

Clazz.newMeth(C$, 'c$$CA$Z',  function (cuts, cutsBefore) {
;C$.$init$.apply(this);
this.cuts=cuts;
this.cutsBefore=cutsBefore;
}, 1);

Clazz.newMeth(C$, 'cut$S',  function (sequence) {
var cutSequences=Clazz.new_($I$(1,1));
sequence=C$.checkSequence$S(sequence);
var chars=sequence.toCharArray$();
var nCut=this.cuts.length;
var p0=0;
var p1=0;
for (var i=0, n=chars.length - 1; i <= n; i++) {
var aa=chars[i];
if (!this.cutsBefore) {
++p1;
}var afterAA=(i < n ? chars[i + 1] : " ");
for (var j=0; j < nCut; j++) {
var c=this.cuts[j];
if (this.shouldCutHere$C$C$C(aa, c, afterAA)) {
C$.makeIon$S$I$I$java_util_ArrayList(sequence, p0, p1, cutSequences);
p0=p1;
}}
if (this.cutsBefore) {
++p1;
}}
if (p1 < sequence.length$()) {
C$.makeIon$S$I$I$java_util_ArrayList(sequence, p1, sequence.length$(), cutSequences);
}return cutSequences;
});

Clazz.newMeth(C$, 'shouldCutHere$C$C$C',  function (aa, c, afterAA) {
return (aa == c);
});

Clazz.newMeth(C$, 'checkSequence$S',  function (sequence) {
if (sequence.contains$CharSequence(" ")) {
throw Clazz.new_(Clazz.load('main.java.Electro2D.ProteaseException').c$$S,["Sequence to be cut must not contain spaces."]);
} else if (sequence.matches$S(".*\\d.*")) {
throw Clazz.new_(Clazz.load('main.java.Electro2D.ProteaseException').c$$S,["Sequence to be cut must not contain numbers."]);
} else if (sequence.matches$S(".*[a-z].*")) {
throw Clazz.new_(Clazz.load('main.java.Electro2D.ProteaseException').c$$S,["Sequence to be cut must contain all upper case letters."]);
}return sequence;
}, 1);

Clazz.newMeth(C$, 'makeIon$S$I$I$java_util_ArrayList',  function (sequence, p0, p1, cutSequence) {
if (p0 < p1) cutSequence.add$O(sequence.substring$I$I(p0, p1));
}, 1);

Clazz.newMeth(C$);
})();
;Clazz.setTVer('3.3.1-v4');//Created 2022-04-21 17:00:07 Java2ScriptVisitor version 3.3.1-v4 net.sf.j2s.core.jar version 3.3.1-v4
